examples/llm/inference/benchmarking/benchmark.py

```diff
@@ -0,0 +1,132 @@
+
+        # Checking if deepspeed casts dtypes correctly
+        for _, p in model.named_parameters():
+            compare_precision(config.precision, p.dtype)
+            break
+    else:
+        model.to(torch.cuda.current_device())
+
+    n_params = sum(p.numel() for p in model.parameters())
+    print('n_params is: ', n_params)
+
+    print('Run name, latency (s), tokens per second')
+    print('=' * 75)
+
+    stats = []
+    for batch_size in config.batch_sizes:
+        for input_length in config.input_lengths:
+            for output_length in config.output_lengths:
+                times = []
+
+                batch = torch.randint(
+                    0,
+                    config.model.vocab_size - 1,
+                    size=(
+                        batch_size,
+                        input_length)).to(f'cuda:{torch.cuda.current_device()}')
+
+                # Make sure we are not generating a fake batch with a EOS token
+                eos_mask = batch == tokenizer.eos_token_id
+                batch = batch.masked_fill(eos_mask,
+                                          config.tokenizer.non_eos_token_id)
+
+                batch = batch.to(torch.long)
+
+                attention_mask = torch.logical_not(
+                    torch.eq(batch, tokenizer.eos_token_id))
+
+                torch.cuda.synchronize()
+
+                for i in range(config.num_runs + 1):
+                    start_time = time.time()
+                    with torch.no_grad():
+                        precision_context = contextlib.nullcontext()
+                        if config.use_precision_context:
+                            precision_context = get_precision_context(
+                                config.precision)
+
+                        with precision_context:
+                            model.generate(batch,
+                                           max_new_tokens=output_length,
+                                           use_cache=True,
+                                           attention_mask=attention_mask)
+
+                    # We noticed there sometimes might be a small bit of startup time
+                    # so we only start to benchmark after some number of batches
+                    if i >= config.num_warmup_batches:
+                        times.append(time.time() - start_time)
+
+                    torch.cuda.synchronize()
+
+                num_output_tokens = output_length * batch_size
+                mean_time = np.mean(times)
+                tokens_per_second = num_output_tokens / float(mean_time)
+
+                result = (
+                    f'{config.benchmark_name}_{batch_size}_{input_length}_{output_length}',
+                    f'{mean_time:.3f}', f'{tokens_per_second:.3f}')
+
+                run_name, latency, tokens_per_second = result
+
+                print(f'{run_name}, {latency}, {tokens_per_second}')
+
+                stats.append(result)
```

Comment:
It might be worth fleshing this out with a couple more values like `ms_per_seq_output_token` which is closer to the relevant user metric "how long do *I* have to wait for eac of *my* tokens when I'm lumped in with everyone else in a batch". Maybe like the rows of the 'inference-perf' google sheet.

And maybe adding strings to the `result` like `f'latency={mean_time:.3f}, ...'` because otherwise it'll look like a bunch of numbers and hard to remember what each one is

Reply:
Ah so right now I kept it display sth like:

Run name, latency (s), tokens per second
===========================================================================
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_1_128_8, 0.081, 99.302
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_2_128_8, 0.081, 196.963
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_4_128_8, 0.082, 392.072
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_8_128_8, 0.087, 731.439
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_16_128_8, 0.100, 1285.682
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_32_128_8, 0.128, 2007.501
MosaicDeepspeed_1xA100-80GB-bf16-GPT-1B_64_128_8, 0.226, 2267.875

to make it easy to copy and paste into the google spreadsheet. Happy to change as well though. 

I will add `ms_per_seq_output_token` btw

Reply:
Actually nvm I see how to amend this. All good!